spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/MaterializingIterator.scala

```diff
@@ -0,0 +1,37 @@
+package com.datastax.spark.connector.util
+
+import java.util.NoSuchElementException
+
+import scala.collection.mutable
+
+/** Materializes/Fetches up to `materializationLimit` from underlying `iterator`.
+  * Materialization happens on `next()` invocation.
+  *
+  * In following example `it.next()` is invoked three times, results are caches in internal queue.
+  * {{{
+  * new MaterializingIterator(it, 3).next()
+  * }}}
+  *
+  */
+class MaterializingIterator[T](iterator: Iterator[T], materializationLimit: Int) extends Iterator[T] {
+
+  assert(materializationLimit > 0, "Materialization limit must be greater than 0")
+
+  private val materialized: mutable.Queue[T] = new mutable.Queue[T]()
+
+  private def materialize(): Unit = {
+    while (iterator.hasNext && materialized.size < materializationLimit) {
+      materialized.enqueue(iterator.next)
+    }
+  }
+
+  override def hasNext = !materialized.isEmpty || iterator.hasNext
+
+  override def next() = {
+    if(!hasNext) {
+      throw new NoSuchElementException("No more elements")
+    }
+    materialize()
```

Comment:
I think in the implementation it's a bit confusing. Won't this basically just equate to executing BUFFER statements async. Then waiting for at least one of them to be "gettable" before moving issuing another? The other issue is if we mix one Long query in with a bunch of slow ones. I think we may be better off just issuing async queries for everything in the iterator at once, and the force get's at the exit of the iterator. This way the Java Driver would be responsible for execution pooling. 

1. Get a ResultFuture for every join query
2. Let the driver figure out when to schedule everything
3. Use map (.get) to force syncing on the way out

I think that the driver and C* are stable enough for this now but I could be wrong